cpp/src/arrow/compute/exec.cc

```diff
@@ -110,6 +112,78 @@ int64_t ExecBatch::TotalBufferSize() const {
   return sum;
 }
 
+bool AddBuffersToSet(const ArrayData& array_data,
+                     std::unordered_set<std::shared_ptr<Buffer>>* seen_buffers) {
```

Comment:
Nit: to reduce the performance overhead, `std::unordered_set<Buffer*>` would be better.